providers/azure.go

```diff
@@ -310,34 +369,84 @@ func makeAzureHeader(accessToken string) http.Header {
 	return makeAuthorizationHeader(tokenTypeBearer, accessToken, nil)
 }
 
-func getEmailFromJSON(json *simplejson.Json) (string, error) {
-	var email string
-	var err error
+func (p *AzureProvider) acquireMicrosoftGraphToken(ctx context.Context) error {
 
-	email, err = json.Get("mail").String()
+	cred, err := confidential.NewCredFromSecret(p.ClientSecret)
+	if err != nil {
+		return err
+	}
+	app, err := confidential.New(p.ClientID, cred, confidential.WithAuthority(p.LoginURL.String()))
+	if err != nil {
+		return err
+	}
 
-	if err != nil || email == "" {
-		otherMails, otherMailsErr := json.Get("otherMails").Array()
-		if len(otherMails) > 0 {
-			email = otherMails[0].(string)
+	var result confidential.AuthResult
+
+	if p.MicrosoftGraphAccessToken == "" {
+		result, err = app.AcquireTokenByCredential(ctx, []string{"openid", "email", "profile", azureGraphURL.String()})
```

Comment:
Would it make more sense to be using the on behalf of flow here instead of client credentials?
Would we want the call to get the groups of a user to come from the user itself? (aka calling `/me`) vs requiring the client registration to have wider reaching permissions to query for all users in AD?

Reply:
I've tried using the on behalf of flow with
```
		result, err = app.AcquireTokenOnBehalfOf(ctx, session.AccessToken, []string{"openid", "email", "profile", azureGraphURL.String()})
```
but obviously it is failing with:
- when passing a resource using `--resource` flag:
```
"error":"invalid_grant","error_description":"AADSTS500131: Assertion audience does not match the Client app presenting the 
assertion. The audience in the assertion was '<resource_ID>' and the expected audience is 
'<client_ID>' or one of the Application Uris of this application with App ID 
'<client_ID>'(<app_name>). The downstream client must request a token for the expected 
audience (the application that made the OBO request) and this application should use that token as the assertion"
```

- without passing the `--resource` flag:
```
"error":"invalid_grant","error_description":"AADSTS50013: Assertion failed signature validation. [Reason - The provided signature 
value did not match the expected signature value."
```

We might need another token with `aud` set for the `client_id`

Reply:
Isn't the token you are getting the first time already one with `aud` set to the `client_id`? Since in v2 endpoints the default is to set the `aud` to `client_id`? I'm just not sure I understand why OBO does not work without the `--resource` flag. I agree I wouldn't expect it to work with `--resource` but that is also a "deprecated" use case on the AAD APIs since it is completely removed in the Microsoft Identity Platform (v2) so I figured we would maybe only support the groups claim on the v2 endpoints where we can use OBO.

Reply:
I can confirm here that using a 7.2.1 branch and the access token issued during flow, you can call /me/getMemberGroups to obtain the groups for a user.

`func (p *AzureProvider) getGroupsFromProfileAPI(ctx context.Context, accessToken string) ([]string, error) {
	if accessToken == "" {
		return nil, errors.New("missing access token")
	}

	groupsURL := p.ProfileURL.String() + "/transitiveMemberOf?&$select=id"

	json, err := requests.New(groupsURL).
		WithContext(ctx).
		WithHeaders(makeAzureHeader(accessToken)).
		Do().
		UnmarshalJSON()
	if err != nil {
		return nil, err
	}

	return getGroupsFromJSON(json)
}

func getGroupsFromJSON(json *simplejson.Json) ([]string, error) {

	var groups []string
	var err error

	values, err := json.Get("value").Array()
	if err != nil || values == nil {
		for _, item := range values {
			if m, ok := item.(map[string]interface{}); ok {
				groups = append(groups, m["id"].(string))
			}

		}
	}

	return groups, err
}`

Reply:
@haydn-j-evans , @braunsonm , you might be right. I'm pretty sure using`/me` was my first attempt a while ago when I started the PR, but not sure why I considered it wasn't working.
Anyway, now that I got back of my long vacation, I took a closer look and it looks like the access token can be used for talking with MS graph, without any extra calls. If you're fine with deprecating `--resource` for v2 then I think we should be fine now. I've adjusted the code accordingly, please give it another try.

Reply:
Any reason we cannot do a token swap once our session is populated? Imo at this point we have no need for the graph token anymore. 

After the session is ready, and if the resource flag is set, could we not then use the refresh token to obtain a new access token with the scope/resource set to the value in the resource flag?

This would also need to be implemented in the cookie refresh logic too.